src/genetic_low_memory.c

```diff
@@ -0,0 +1,459 @@
+  char secondChildString[encodedChildStringSize];
+
+  firstChildString[arraySize] = '\0';
+  secondChildString[arraySize] = '\0';
+
+  for (unsigned int i = 0; i < (UINT16_DIGIT_COUNT * dimensions); i++) {
+
+    if (linear_congruential_random_generator() <= 0.5) {
+
+      firstChildString[i] = firstParent[i];
+      secondChildString[i] = secondParent[i];
+    } else {
+      firstChildString[i] = secondParent[i];
+      secondChildString[i] = firstParent[i];
+    }
+  }
+
+  // We apply the mutation operator to both children
+  mutate(firstChildString, arraySize);
+  mutate(secondChildString, arraySize);
+
+  // We decode both children and add them to the next generation
+  decodeAndAddChild(nextGeneration, nextGenerationSize, firstChildString);
+  decodeAndAddChild(nextGeneration, nextGenerationSize, secondChildString);
+}
+
+/**
+ * We encode a solution as a string , the given string will always be the same
+ * length
+ *
+ * @param combinedValue the created string containing the encoded parent
+ * @param parent an array containing the parameters of the parent
+ */
+static void encode(char* combinedValue, uint16_t* parent) {
+
+  unsigned int arrIndex = 0;
+  unsigned int encodedParameterSize = UINT16_DIGIT_COUNT + 1;
+
+  for (unsigned int i = 0; i < dimensions; i++) {
+
+    char stringValue[encodedParameterSize];
+    char otherString[encodedParameterSize];
+
+    stringValue[UINT16_DIGIT_COUNT] = '\0';
+    otherString[UINT16_DIGIT_COUNT] = '\0';
+
+    unsigned int otherStringLength = 0;
+    for (int j = (UINT16_DIGIT_COUNT - 1); j >= 0; j--) {
+
+      if (parent[i] < pow(10, j)) {
+        otherString[otherStringLength] = '0';
+        otherStringLength++;
+      } else {
+        break;
+      }
+    }
+
+    sprintf(stringValue, "%hu", parent[i]);
+
+    if (otherStringLength != 0) {
+      for (unsigned int j = 0; j < otherStringLength; j++) {
+        combinedValue[j + arrIndex] = otherString[j];
+      }
+    }
+
+    for (unsigned int j = otherStringLength; j < UINT16_DIGIT_COUNT; j++) {
+
+      combinedValue[j + arrIndex] = stringValue[j - otherStringLength];
+    }
+    arrIndex += UINT16_DIGIT_COUNT;
+  }
+}
+
+/**
+ * This function takes care of creating the next generation and selecting the
+ * parents
+ *
+ * @param population  this array stores all the values of the population
+ * @param nextGeneration this array is used to store the parametres of the
+ * created children
+ * @param evaluationFunction the function that is used to evaluate each solution
+ * in the population
+ */
+static void createNextGeneration(uint16_t* population, uint16_t* nextGeneration,
+
+                                 float (*evaluationFunction)(float*)) {
+
+  unsigned int currentNextGenerationSize = 2;
+  const unsigned int mergedParentsLength =
+      (dimensions * UINT16_DIGIT_COUNT) + 1;
+  while (currentNextGenerationSize < (populationSize)) {
+
+    unsigned int parent1Index, parent2Index = 0;
```

Comment:
You are assigning to value to parent2Index but not parent1Index, is this the expected behaviour ?